core/src/main/scala/kafka/cluster/Replica.scala

```diff
@@ -36,25 +36,51 @@ class Replica(val brokerId: Int,
   @volatile private[this] var logEndOffsetMetadata: LogOffsetMetadata = LogOffsetMetadata.UnknownOffsetMetadata
 
+  // The log end offset value at the time the leader receives last FetchRequest from this follower.
+  // This is used to determine the last catch-up time of the follower
+  @volatile var lastFetchLeaderLogEndOffset: Long = 0L
+
+  // The time when the leader receives last FetchRequest from this follower
+  // This is used to determine the last catch-up time of the follower
+  @volatile var lastFetchTimeMs: Long = 0L
+
   val topic = partition.topic
   val partitionId = partition.partitionId
 
   def isLocal: Boolean = log.isDefined
 
-  private[this] val lastCaughtUpTimeMsUnderlying = new AtomicLong(time.milliseconds)
+  // lastCatchUpTimeMs is the largest time t such that the begin offset of most recent FetchRequest from this follower >=
+  // the LEO of leader at time t. This is used to determine the lag of this follower and ISR of this partition.
+  private[this] val lastCaughtUpTimeMsUnderlying = new AtomicLong(0L)
 
   def lastCaughtUpTimeMs = lastCaughtUpTimeMsUnderlying.get()
 
+  /*
+   * If the FetchRequest reads up to the log end offset of the leader when the current fetch request was received,
+   * set the lastCaughtUpTimeMsUnderlying to the time when the current fetch request was received.
+   *
+   * Else if the FetchRequest reads up to the log end offset of the the leader when the previous fetch request was received,
+   * set the lastCaughtUpTimeMsUnderlying to the time when the previous fetch request was received.
+   *
+   * This is needed to enforce the semantics of ISR, i.e. a replica is in ISR if and only if it lags behind leader's LEO
+   * by at most replicaLagTimeMaxMs. This semantics allows a follower to be added to the ISR even if offset of its fetch request is
+   * always smaller than leader's LEO, which can happen if there are constant small produce requests at high frequency.
+   */
   def updateLogReadResult(logReadResult : LogReadResult) {
+    if (logReadResult.info.fetchOffsetMetadata.messageOffset >= logReadResult.leaderLogEndOffset)
+      lastCaughtUpTimeMsUnderlying.set(logReadResult.fetchTimeMs)
+    else if (logReadResult.info.fetchOffsetMetadata.messageOffset >= lastFetchLeaderLogEndOffset)
+      lastCaughtUpTimeMsUnderlying.set(lastFetchTimeMs)
+
     logEndOffset = logReadResult.info.fetchOffsetMetadata
+    lastFetchLeaderLogEndOffset = logReadResult.leaderLogEndOffset
+    lastFetchTimeMs = logReadResult.fetchTimeMs
+  }
 
-    /* If the request read up to the log end offset snapshot when the read was initiated,
-     * set the lastCaughtUpTimeMsUnderlying to the current time.
-     * This means that the replica is fully caught up.
-     */
-    if(logReadResult.isReadFromLogEnd) {
-      lastCaughtUpTimeMsUnderlying.set(time.milliseconds)
-    }
+  def resetLastCatchUpTime() {
+    lastFetchLeaderLogEndOffset = 0L
+    lastFetchTimeMs = 0L
+    lastCaughtUpTimeMsUnderlying.set(0L)
```

Comment:
Hmm, when the leadership switches to a replica, this could cause an in-sync replica to be out of ISR immediately before it can make a fetch request to the new leader. In this case, we want to give every ISR the benefit of doubt and assume that it's caught up at this moment. For non-in-sync replicas, we want to assume that initially it's not caught up.